src/iterator.c

```diff
@@ -2087,9 +2127,33 @@ static int index_iterator_advance(
 	const git_index_entry *entry = NULL;
 	bool is_submodule;
 	int error = 0;
+	git_index_entry key;
+	git_buf buf = GIT_BUF_INIT;
 
 	iter->base.flags |= GIT_ITERATOR_FIRST_ACCESS;
 
+	if (iter->end_idx == (size_t)-1) {
```

Comment:
The current code performs a linear scan over the index until it finds an entry with path >= `iter->base.start`. This uses up a lot of CPU when checking the status of a single file or when using sharding. This linear scan negatively affects scaling afforded by sharding because each shard has to scan O(N) entries in the index where N is the total size of the index. I changed it so that the first entry is found with binary search.

If pathspec is set, the code will still do linear search between elements of the pathspec even though this could also be done with binary search. So if you want to diff just two files -- `"a"` and `"z"` -- the code will do a binary search to find `"a"` and then scan index linearly until it finds `"z"`.

I didn't optimize this part because I have a workaround in my code. I just issue two diff calls -- one for `"a"` and another for `"z"`.